Src/FluentAssertions/Common/TypeExtensions.cs

```diff
@@ -584,12 +584,29 @@ private static bool IsAnonymousType(this Type type)
     public static bool IsRecord(this Type type)
     {
         return TypeIsRecordCache.GetOrAdd(type, static t =>
-            t.GetMethod("<Clone>$") is not null &&
-            t.GetTypeInfo()
-                 .DeclaredProperties
-                 .FirstOrDefault(p => p.Name == "EqualityContract")?
-                 .GetMethod?
-                 .GetCustomAttribute(typeof(CompilerGeneratedAttribute)) is not null);
+        {
+            var isRecord = t.GetMethod("<Clone>$") is not null &&
+                           t.GetTypeInfo()
+                               .DeclaredProperties
+                               .FirstOrDefault(p => p.Name == "EqualityContract")?
+                               .GetMethod?
+                               .GetCustomAttribute(typeof(CompilerGeneratedAttribute)) is not null;
+
+            // As noted here: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs#open-questions
+            // recognizing record structs from metadata is an open point. The following check is based on common sense
+            // and heuristic testing, apparently giving good results but not supported by official documentation.
+            var isRecordStruct = t.BaseType == typeof(ValueType) &&
+                                 t.GetMethods()
+                                     .Where(m => m.Name == "op_Inequality")
+                                     .SelectMany(m => m.GetCustomAttributes(typeof(CompilerGeneratedAttribute)))
+                                     .Any() &&
+                                 t.GetMethods()
+                                     .Where(m => m.Name == "op_Equality")
+                                     .SelectMany(m => m.GetCustomAttributes(typeof(CompilerGeneratedAttribute)))
+                                     .Any();
```

Comment:
While this is probably the best we can currently do, I wonder a bit if this is still too broad a check?

Reply:
I'm afraid of this too. This is backed by MSDN too (see comment). I've tried my best to "crack" this test and was not able to, but please feel free to add counter-examples.

Reply:
Checking for one operator should be enough.
```cs
var isRecordStruct = t.BaseType == typeof(ValueType) &&
                     t.GetMethod("op_Equality", BindingFlags.Static | BindingFlags.Public | BindingFlags.DeclaredOnly, null, new[] { t, t }, null)?
                        .GetCustomAttribute(typeof(CompilerGeneratedAttribute)) is not null;
```

Reply:
I had another look at creating a heuristic to identify record structs with few false positives.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs#printing-members-printmembers-and-tostring-methods

specifies that a `record struct` contains a synthesized (read has `[CompilerGenerated]`)
```cs
private bool PrintMembers(System.Text.StringBuilder builder)
```

The name `PrintMembers` is also a [public part of Roslyn](https://github.com/dotnet/roslyn/blob/d71b443067e30dbd935a8cffdacda2ca7ba32e7f/src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs#L387-L390).

```cs
var isRecordStruct = t.BaseType == typeof(ValueType) &&
                     t.GetMethod("PrintMembers", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly, null, new[] { typeof(StringBuilder) }, null)?
                        .GetCustomAttribute(typeof(CompilerGeneratedAttribute)) is { };
```

What do you think about this?

Reply:
That  may be a good catch. I'll look into it!

Reply:
@jnyrup that looked promising, but unfortunately you can "crack" the check for a compiler generated PrintMembers, because the compiler does allow you to define your PrintMembers methods with the very same signature in the body of the record struct. I know that this is very unlikely, but nonetheless it is possible.
On the other hand, the compiler does not allow to define your operator==, and to my best knowledge record structs are (currently!) the only value types having a compiler generated equality operator, so I'd go for it. I have used the simplified check you have proposed.

Reply:
Not sure I understand you here 🤔 
As a record struct currently doesn't have any unspeakable parts, it's _always_ possible to write a struct in plain C# that looks like a record struct.

E.g. this is valid C# code, [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAlAVwDsAXASwFsZCBhCWgB2oBsYoBlbgN2pgYAZwDcAWABQ0jAGYcIylFJhKOAILSA3tJz6cAbWZtO3AOIxy3AIaUYAEwC6eg/Lz4kOYBAgccEKy2lNA4ALxhABQaOAgANJo4AJ4AlOEAfDgAZjYcIjCSUgY4rvrGLOxcUJbWUHaOLkVuCgRePn4BQXUhUDgAhFEx8YmpGdm5+WIlUgC+QA===)
```cs
public struct A
{
    [CompilerGenerated]
    public static bool operator ==(A x, A y) => true;
    
    [CompilerGenerated]
    public static bool operator !=(A x, A y) => false 
}
```

I tried another thing.

I made a `record struct Auto {}` and copied its output back into regular C# code.
The _only_ thing I wasn't allowed to was using `[IsReadOnly]` as that gave a compiler error.

```
error CS8335: Do not use 'System.Runtime.CompilerServices.IsReadOnlyAttribute'. This is reserved for compiler usage.
```

`PrintMembers` is annotated with `[IsReadOnly]`

So what do you think about checking for this?

```cs
var isRecordStruct = t.BaseType == typeof(ValueType) &&
                     t.GetMethod("PrintMembers", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly, null, new[] { typeof(StringBuilder) }, null)?
                        .GetCustomAttribute(typeof(IsReadOnlyAttribute)) is { };
```

I don't know if an arbitrary source generator is allowed to use `[IsReadOnly]`.
IL weaving would most likely be able to circumvent CS8335, but we aren't looking for a bullet proof solution.

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0ASgK4B2ALgJYC2MlAwhKwA7MAbGFADKwgG7MwMAM4BuQiXJU6TNhwCSTYRF5iok6fMVkKlACowEjBQUKxIUACakZjKLTCNSAQVqMILABvAF9CQgBtEXdPRgAZAEMATwh/AApElP8AaWZ6J0oxAEdaGDUEgQAaUhFmAC8YUgBeUhwASgBdQjzGYXoK1xivX38IUhBSDQBREoTGBOAhAB4/AIA+QiDCUh3SAHo90giNGWoYBKcAeXoBJK6CXaPuPkFhAHEy4TmYJ3vHjAAzKQIOJhFBmE5GhQyOYINFwfQAOZpNrbXZbB6PXbwvKIgBCtEEkKgymIBKJwmapHoMAA7jV3LjyQJiSjbFjdtDmcTKD5eLwyk40gAiVYQYVtdkc0nc4S8/mCkWkIKkCVSjnMABmpDSAAUEYwALIwVjAYQyNJcwks4RtVGYjkY6VYq0UqDygX5NIAclI3slaI5YQdLpwZOtPL5nqF3pC/vVLoA7DKI3LYTikWzA6Rg9mDsdTucrjc7tmIs9+EIoB8aVBvr9s7xweJvqRgBAIAJSPqesbTea0hn8amScAR/bHk7pRhk5qKjIYAncyHyzxK+9PnXeg2Q4DlEg2x2u7ovgESQBCJppMWkISaxjVG/gxEAC0YE/R2f+yfPaTv3iaFpnzfAMQ2XR5VxeKsa1PH4/k5IEKAPdtO2BAUt2gZorxvf9H1GUhgPfbMpw5Gdbxge9KBmWh5zSQjQMecDdnzE4zgua5bngnZIPXatN3rLjyCBEEwQhRoelID5GAACQSGQX24SEUWIr9OWTYgl3CEMWMLdiS0EnjXj42sBOzPcRKgcFIUPVDqNoiBgAAKxgYYHMcj8dhIrEtR1NzSGYGQRgCDysS850yLsgQLWvUY2jchjpWDcLZ3nRdsyYnYdLY4tOLLCsjJgrc4LMoEUK7SKLRvCBGBfW0VJDb9SBiNKwMIEIgA==)

Reply:
Interesting! I didn't know one could annotate with [CompilerGenerated] in source code; it looks a little silly to me that you can, but it is what it is I suppose :)
In this case, I definitely like your proposal with PrintMembers and IsReadOnly more. Let me try to give a couple of final touches and I'll be back with a new commit.

Reply:
@jnyrup I have just tried the last proposal checking for a PrintMembers annotated with IsReadOnly. IsReadOnly is not available on NET47 and NETSTANDARD2_0, so it does not compile on those target, but here the problem is just my ignorance on properly handling multi-target.
The major problem for me is that the check fails for the MyReadOnlyRecordStruct and MyRecordStructWithCustomPrintMembers tests, for which PrintMembers has not the IsReadOnly attribute.
For this reason, at the end of the day (literally, for me 😄 ) I believe the lesser evil for now is checking for compiler generated equality, knowing that we can have a false positive if someone uses the [CompilerGenerated] attribute in source code (I have added a test to document this).

Reply:
Aha!
I hadn't thought about the case in `MyRecordStructWithCustomPrintMembers` where one implements `PrintMembers` themselves such that the compiler _doesn't_ create one.
Well found 👍 

So it seems there will always be a `PrintMembers`, but we don't know who wrote it.

This code handles all your test cases and the false positive.

```cs
public static bool IsRecord(this Type type)
{
    return TypeIsRecordCache.GetOrAdd(type, static t => t.IsRecordClass() || t.IsRecordStruct());
}

private static bool IsRecordClass(this Type type)
{
    return type.GetMethod("<Clone>$", BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly) is { } &&
            type.GetProperty("EqualityContract", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly)?
                .GetMethod?.IsDecoratedWith<CompilerGeneratedAttribute>() == true;
}

private static bool IsRecordStruct(this Type type)
{
    // As noted here: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs#open-questions
    // recognizing record structs from metadata is an open point. The following check is based on common sense
    // and heuristic testing, apparently giving good results but not supported by official documentation.
    return type.BaseType == typeof(ValueType) &&
            type.GetMethod("PrintMembers", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly, null, new[] { typeof(StringBuilder) }, null) is { } &&
            type.GetMethod("op_Equality", BindingFlags.Static | BindingFlags.Public | BindingFlags.DeclaredOnly, null, new[] { type, type }, null)?
                .IsDecoratedWith<CompilerGeneratedAttribute>() == true;
}
```

I would be confident enough with this one.

Reply:
Good idea: this way we rely on the two documented behaviors of record structs!
There is still a small possibility of a false positive, where one both implements a PrintMember and annotates equality with [CompilerGenerated] (I have added a test for reference), but I'd consider such code as deliberate sabotage 😄